src/System.IO.FileSystem/src/System/IO/PathHelpers.cs

```diff
@@ -38,5 +41,146 @@ internal static bool EndsInDirectorySeparator(string path)
+        /// Combines three paths. Does no validation of paths, only concatenates the paths
+        /// and places a directory separator between them if needed.
+        /// </summary>
+        internal static string CombineNoChecks(string first, ReadOnlySpan<char> second, ReadOnlySpan<char> third)
+        {
+            if (string.IsNullOrEmpty(first))
+                return CombineNoChecks(second, third);
+
+            if (second.Length == 0)
+                return CombineNoChecks(first, third);
+
+            if (third.Length == 0)
+                return CombineNoChecks(first, second);
+
+            return CombineNoChecksInternal(first.AsReadOnlySpan(), second, third);
+        }
+
+        /// <summary>
+        /// Combines two paths. Does no validation of paths, only concatenates the paths
+        /// and places a directory separator between them if needed.
+        /// </summary>
+        internal unsafe static string CombineNoChecks(string first, string second)
+        {
+            if (string.IsNullOrEmpty(first))
+                return string.IsNullOrEmpty(second) ? string.Empty : second;
+
+            if (string.IsNullOrEmpty(second))
+                return first;
+
+            return CombineNoChecksInternal(first.AsReadOnlySpan(), second.AsReadOnlySpan());
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second)
+        {
+            Debug.Assert(first.Length > 0 && second.Length > 0, "should have dealt with empty paths");
+
+            bool hasSeparator = first[first.Length - 1] == Path.DirectorySeparatorChar
+                || second[0] == Path.DirectorySeparatorChar;
+
+            int totalLength = first.Length + second.Length + (hasSeparator ? 0 : 1);
+            string fullPath = new string('\0', totalLength);
+            fixed (char* f = fullPath)
+            {
+                Span<char> pathSpan = new Span<char>(f, totalLength);
+                first.CopyTo(pathSpan);
+                if (!hasSeparator)
+                    pathSpan[first.Length] = Path.DirectorySeparatorChar;
+                second.CopyTo(pathSpan.Slice(first.Length + (hasSeparator ? 0 : 1)));
+            }
+            return fullPath;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second, ReadOnlySpan<char> third)
+        {
+            Debug.Assert(first.Length > 0 && second.Length > 0 && third.Length > 0, "should have dealt with empty paths");
+
+            bool firstHasSeparator = first[first.Length - 1] == Path.DirectorySeparatorChar
+                || second[0] == Path.DirectorySeparatorChar;
+            bool thirdHasSeparator = second[second.Length - 1] == Path.DirectorySeparatorChar
+                || third[0] == Path.DirectorySeparatorChar;
```

Comment:
Some places we're assuming Path.DirectorySeparatorChar and other places we're checking Alt as well with PathInternal.IsDirectorySeparator. Can we assume Windows will always give us Path.DirectorySeparatorChar and not use PathInternal.IsDirectorySeparator. ?

Reply:
It is worth reviewing. Need to do some cleanup here.